relay-server/src/utils/dynamic_sampling.rs

```diff
@@ -586,10 +807,122 @@ mod tests {
 
         for (rule_test_name, condition) in conditions.iter() {
             let failure_name = format!("Failed on test: '{}'!!!", rule_test_name);
-            assert!(condition.matches(&tc), failure_name);
+            assert!(condition.matches_trace(&tc, None), failure_name);
         }
     }
 
+    #[test]
+    /// test matching for various rules
+    fn test_matches_events() {
```

Comment:
In a follow-up, I would like to move all these tests to macro-based expansion. That way, each test runs in its own test function and we can see individual failures more easily.

We could also start adopting a test parameterization framework, but I'm still not convinced of that.

Reply:
While I like the idea in principle it would make debugging difficult/impossible and I like to have this option open. 
I've been looking, a while back now, for a clean way to do parameterization of unit tests in Rust but all I found were frameworks that either used nightly builds or used some ugly macros.

If we can find a relatively clean way to do it while still being able to run a debugger I all for it.

Reply:
You can run and debug tests defined in macros just like any other test. Rust Analyzer even provides code lenses for debugging individual tests, for example as we do in [`relay-general/tests/test_fixtures.rs`](https://github.com/getsentry/relay/blob/569678659bac656d6c4a6edcbdf3381677b5eaa2/relay-general/tests/test_fixtures.rs#L102-L107):

![image](https://user-images.githubusercontent.com/1433023/109279035-8f61d100-7819-11eb-870f-c956f3689af9.png)